src/java.desktop/share/classes/com/sun/java/swing/SwingUtilities3.java

```diff
@@ -135,4 +143,73 @@ public static RepaintManager getDelegateRepaintManager(Component
         }
         return delegate;
     }
+
+    @FunctionalInterface
+    public interface PaintInterface {
+        void paintUnscaledBorder(Component c, Graphics g, int x, int y,
+                                 int w, int h, double scale, int strokeWidth);
+    }
+
+    public static void paintBorder(Component c, Graphics g, int x, int y,
+                            int w, int h, PaintInterface paintFunction) {
+
+        //STEP 1: RESET TRANSFORM
+        //save old values
+        AffineTransform at = null;
+        Stroke oldStk = null;
+        boolean resetTransform = false;
+        int stkWidth = 1;
+        double scaleFactor = 1;
+
+        if (g instanceof Graphics2D g2d) {
+            at = g2d.getTransform();
+            scaleFactor = at.getScaleX();
+            oldStk = g2d.getStroke();
+
+            // if m01 or m10 is non-zero, then there is a rotation or shear
+            // skip resetting the transform
+            resetTransform = ((at.getShearX() == 0) && (at.getShearY() == 0));
+
+            if (resetTransform) {
+                g2d.setTransform(new AffineTransform());
+                if (c instanceof JInternalFrame) {
+                    stkWidth = clipRound(Math.min(at.getScaleX(), at.getScaleY()));
+                } else {
+                    stkWidth = (int) Math.floor(Math.min(at.getScaleX(), at.getScaleY()));
+                }
+                g2d.setStroke(new BasicStroke((float) stkWidth));
+            }
+        }
+
+        int xtranslation = 0;
+        int ytranslation = 0;
+        int width = 0;
+        int height = 0;
+
+        if (resetTransform) {
+            double xx = at.getScaleX() * x + at.getTranslateX();
```

Comment:
Can we combine this code block with the above if (resetTransform) at line 173 ? or did you keep it separate for a reason?

Reply:
@rajamah In-order to paint the border even when g is not a Graphics2D object I have added the resetTransform check here again and additionally to keep the if & else block close-by, which is easier to understand rather than having the blocks separated.

Reply:
sure, makes sense. Thanks

Reply:
Still, I think it makes sense to combine this block with the above one.

If `resetTransform` is true, the transform is reset and variables are set to the values as calculated in this block; otherwise, the variables are initialised with the default values.